File: java/alldifferentdirections.java
import java.util.Scanner;

public class alldifferentdirections {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int n;
		while ((n = scan.nextInt()) != 0) {
			scan.nextLine();
			Point[] pt_list = new Point[n];
			for (int i = 0; i < n; i++) {
				String[] dir_input = scan.nextLine().split(" ");
				Point pt = new Point(Double.parseDouble(dir_input[0]), Double.parseDouble(dir_input[1]));
				for (int j = 2; j < dir_input.length; j += 2) {
					double unit = Double.parseDouble(dir_input[j + 1]);
					switch (dir_input[j]) {
					case "start":
						pt.setDirection(unit);
						break;
					case "turn":
						pt.turnDirection(unit);
						break;
					default:
						pt.walk(unit);
						break;
					}
				}
				pt_list[i] = pt;
			}
			double x = 0.0;
			double y = 0.0;
			for (Point p : pt_list) {
				x += p.x;
				y += p.y;
			}
			Point avg_pt = new Point(x / pt_list.length, y / pt_list.length);
			double worse = 0.0;
			for (Point p : pt_list)
				worse = Math.max(worse, avg_pt.getDistance(p));
			System.out.printf("%f %f %f%n", avg_pt.x, avg_pt.y, worse);
		}
		scan.close();
	}
}

class Point {
	public double x, y;
	double dir;

	public Point(double x, double y) {
		this.x = x;
		this.y = y;
	}

	public void setDirection(double dir) {
		this.dir = dir;
		this.dir %= 360;
		if (this.dir < 0)
			this.dir += 360;
	}

	public void turnDirection(double dir) {
		setDirection(this.dir + dir);
	}

	public void walk(double unit) {
		x += unit * Math.cos(dir * Math.PI / 180);
		y += unit * Math.sin(dir * Math.PI / 180);
	}

	public double getDistance(Point pt) {
		double X_d = pt.x - x;
		double Y_d = pt.y - y;
		return Math.sqrt(X_d * X_d + Y_d * Y_d);
	}
}

File: ts/src/alldifferentdirections.ts
import readline = require('readline')

function getFinalPoint (input: string[]) {
  const p = new Point(parseFloat(input[0] ?? '0'), parseFloat(input[1] ?? '0'))
  for (let i = 2; i < input.length; i += 2) {
    const value = parseFloat(input[i + 1] ?? '0')
    switch (input[i]) {
      case 'start':
      case 'turn':
        p.turn(value)
        break
      default:
        p.walk(value)
        break
    }
  }
  return p
}

class Point {
  private degree: number

  constructor (public x = 0, public y = 0) {
    this.degree = 0
    this.x = x
    this.y = y
  }

  turn (degrees: number) {
    this.degree += degrees
  }

  walk (distance: number) {
    this.x += Math.cos((this.degree * Math.PI) / 180) * distance
    this.y += Math.sin((this.degree * Math.PI) / 180) * distance
  }

  static averagePoint (points: Point[]) {
    const sumPoint = points.reduce((p1, p2) => {
      return new Point(p1.x + p2.x, p1.y + p2.y)
    }, new Point(0, 0))
    return new Point(sumPoint.x / points.length, sumPoint.y / points.length)
  }

  static getDistance (p1: Point, p2: Point) {
    return Math.sqrt(Math.pow(p1.x - p2.x, 2) + Math.pow(p1.y - p2.y, 2))
  }

  static worstDistance (points: Point[], point: Point) {
    return Math.max(
      ...points.map((p) => {
        return Point.getDistance(p, point)
      })
    )
  }
}
const regNumbersOnly = /^\d+$/
let dirs = 0
let finalPoints: Point[] = []

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line) => {
    if (regNumbersOnly.test(line)) {
      if (finalPoints.length > 0) {
        const averagePoint = Point.averagePoint(finalPoints)
        const worstDistance = Point.worstDistance(finalPoints, averagePoint)
        console.log(`${averagePoint.x.toFixed(4)} ${averagePoint.y.toFixed(4)} ${worstDistance.toFixed(4)}`)
        finalPoints = []
      }
      dirs = parseInt(line)
      if (dirs === 0) {
        process.exit(0)
      }
    } else {
      const point = getFinalPoint(line.split(' '))
      finalPoints.push(point)
    }
  })
  .on('close', () => {
    process.exit(0)
  })
